src/Orleans.Serialization.ProtobufNet/ProtobufNetSerializer.cs

```diff
@@ -0,0 +1,105 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.IO;
+using ProtoBuf;
+
+namespace Orleans.Serialization.ProtobufNet
+{
+    /// <summary>
+    /// An implementation of IExternalSerializer for usage with Protobuf types, using the protobuf-net library.
+    /// </summary>
+    public class ProtobufNetSerializer : IExternalSerializer
+    {
+        private static readonly ConcurrentDictionary<RuntimeTypeHandle, bool> IsSupported = new ConcurrentDictionary<RuntimeTypeHandle, bool>();
+
+        /// <summary>
+        /// Determines whether this serializer has the ability to serialize a particular type.
+        /// </summary>
+        /// <param name="itemType">The type of the item to be serialized</param>
+        /// <returns>A value indicating whether the type can be serialized</returns>
+        public bool IsSupportedType(Type itemType)
+        {
+            if (IsSupported.TryGetValue(itemType.TypeHandle, out var isSupported))
+            {
+                return isSupported;
+            }
+            isSupported = Attribute.GetCustomAttribute(itemType, typeof(ProtoContractAttribute)) != null;
+            IsSupported.AddOrUpdate(itemType.TypeHandle, isSupported, (type, val) => isSupported);
+            return isSupported;
+        }
+
+        /// <inheritdoc />
+        public object DeepCopy(object source, ICopyContext context)
+        {
+            if (source == null)
+            {
+                return null;
+            }
+
+            dynamic dynamicSource = source;
+            return dynamicSource.Clone();
+        }
+
+        /// <inheritdoc />
+        public void Serialize(object item, ISerializationContext context, Type expectedType)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (item == null)
+            {
+                // Special handling for null value. 
+                // Since in this ProtobufSerializer we are usually writing the data lengh as 4 bytes
+                // we also have to write the Null object as 4 bytes lengh of zero.
+                context.StreamWriter.Write(0);
+                return;
+            }
+            
+            using (var stream = new MemoryStream())
+            {
+                ProtoBuf.Serializer.Serialize(stream, item);
```

Comment:
I wonder if users would ever want to use a `RuntimeTypeModel` other than `RuntimeTypeModel.Default` and in that case whether or not it makes sense to pass that in via some options. I don't know for sure, since I'm not very experienced with protobuf-net.